modules/Bio/EnsEMBL/Compara/Utils/ReferenceDatabase.pm

```diff
@@ -0,0 +1,224 @@
+            }
+            my $msg = "\n\n** Reference GenomeDB with this name [$species_production_name], assembly".
+            " [$this_assembly] and genebuild.last_geneset_update [$this_genebuild] is already in the compara DB **\n";
+            if ($match) {
+                $msg .= "** And it has the right set of DnaFrags **\n";
+                $msg .= "** You can use the --force option to update the other GenomeDB fields IF YOU REALLY NEED TO!! **\n\n";
+            } else {
+                $msg .= "** But the DnaFrags don't match: **\n$output\n";
+                $msg .= "** You can use the --force option to update the DnaFrag, but only IF YOU REALLY KNOW WHAT YOU ARE DOING!! **\n\n";
+            }
+            throw $msg;
+        }
+        else {
+            print "Reference GenomeDB before update: ", $stored_genome_db->toString, "\n";
+
+            # Get fresher information from the core database
+            $stored_genome_db->db_adaptor($species_dba, 1);
+            $stored_genome_db->last_release(undef);
+
+            # And store it back in Compara
+            $genome_db_adaptor->update($stored_genome_db);
+            $genome_db = $stored_genome_db;
+        }
+    } else { # new genome or new assembly!!
+        $new_genome_db->taxon_id( $taxon_id ) if $taxon_id;
+
+        if (!defined($new_genome_db->name)) {
+            throw "Cannot find species.production_name in meta table for ".($species_dba->locator).".\n";
+        }
+        if (!defined($new_genome_db->taxon_id)) {
+            throw "Cannot find species.taxonomy_id in meta table for ".($species_dba->locator).".\n".
+                  "   You can use the --taxon_id option";
+        }
+        print "New reference GenomeDB for Compara: ", $new_genome_db->toString, "\n";
+
+        # new ID search if $offset is true
+        if($offset) {
+            my ($max_id) = $compara_dba->dbc->db_handle->selectrow_array('select max(genome_db_id) from genome_db where genome_db_id > ?', undef, $offset);
+            $max_id = $offset unless $max_id;
+            $new_genome_db->dbID($max_id + 1);
+        }
+        $genome_db_adaptor->store($new_genome_db);
+        $genome_db = $new_genome_db;
+    }
+
+    $genome_db_adaptor->make_object_current($genome_db);
+    return $genome_db;
+}
+
+
+############################################################
+#          remove_reference_genome.pl methods              #
+############################################################
+
+=head2 remove_reference_genome
+
+  Arg[1]      : Bio::EnsEMBL::Compara::DBSQL::DBAdaptor $compara_dba
+  Arg[2]      : Bio::EnsEMBL::Compara::GenomeDB $genome_db
+  Description : Remove all trace of $genome_db from the $compara_dba database. This includes
+                dnafrags and members
+  Returns     : none
+  Exceptions  : throws if $genome_db is not a Bio::EnsEMBL::Compara::GenomeDB
+
+=cut
+
+sub remove_reference_genome {
```

Comment:
There is no `GenomeDB->delete` method, mostly because it's not really something we do - in the current system, we only ever add to the master, never delete. And we populate other dbs from that using only the genomes it needs, so have never needed to delete there either.

The second issue with what you're proposing is how to implement it. For example, if we were to call `delete` on a genome that is in an EPO alignment, wouldn't we need to remove it from all blocks, minimize the blocks&tree, etc (basically what you've been working on for a while now). And that's only considering one type of data!

So essentially, I think it's a bit more work than required to properly implement a `GenomeDB->delete` function. At least with this `remove_reference_genome`, it's clear from the name that it's just designed to work on a reference master, where the types of data linked to the genome_db is a controlled/known entity.